cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -925,14 +943,32 @@ struct BatchConverter {
         });
   }
 
+  Result<std::shared_ptr<Schema>> InitializeSchema(
+      const std::vector<std::string>& names) {
+    // By this point this->schema will have been set by the SinkNode.  We potentially
+    // rename it with the names provided by the user and then return this in case the user
+    // wants to know the output schema.
+    if (!names.empty()) {
+      if (static_cast<int>(names.size()) != schema->num_fields()) {
+        return Status::Invalid(
+            "A plan was created with custom field names but the number of names did not "
+            "match the number of output columns");
```

Comment:
nit: include expected/actual count in the message for convenience?

Reply:
Done.